File: whisper_align.py
#!/usr/bin/env python3
"""
Use Whisper to automatically detect lyric timing
Usage: python3 whisper_align.py <audio-file> <lyrics-file> <output-dir>
"""

import sys
import json
import os
from pathlib import Path

def main():
    if len(sys.argv) < 3:
        print('Usage: python3 whisper_align.py <audio-file> <lyrics-file> [output-dir]')
        print('\nExample:')
        print('  python3 whisper_align.py song.mp3 lyrics.txt ./output')
        sys.exit(1)
    
    audio_file = sys.argv[1]
    lyrics_file = sys.argv[2]
    output_dir = sys.argv[3] if len(sys.argv) > 3 else './output'
    
    print('🎵 Whisper Automatic Alignment\n')
    
    # Import whisper (will download model if needed)
    print('📦 Loading Whisper...')
    try:
        import whisper
        import ssl
        ssl._create_default_https_context = ssl._create_unverified_context
    except ImportError:
        print('❌ Whisper not installed. Install with: pip3 install openai-whisper')
        sys.exit(1)
    
    # Read lyrics
    print('📖 Reading lyrics...')
    with open(lyrics_file, 'r', encoding='utf-8') as f:
        lyrics = [line.strip() for line in f if line.strip()]
    
    print(f'   Found {len(lyrics)} lines\n')
    
    # Load Whisper model
    print('🎤 Loading Whisper model (first time will download ~150MB)...')
    model = whisper.load_model("base")
    
    # Transcribe with word timestamps
    print('🎤 Analyzing audio (this may take 2-5 minutes)...')
    print('   Detecting speech and timing...')
    
    result = model.transcribe(
        audio_file,
        language='hi',
        word_timestamps=True,
        verbose=False
    )
    
    print('   ✓ Whisper analysis complete\n')
    
    # Match lyrics to segments
    print('🔗 Matching lyrics to detected speech...')
    timed_lyrics = match_lyrics_to_segments(lyrics, result['segments'])
    
    # Create timestamps.json
    total_duration = timed_lyrics[-1]['endTime'] if timed_lyrics else 0
    
    data = {
        'version': '1.0',
        'metadata': {
            'generatedAt': '2025-12-05T22:00:00.000Z',
            'totalLines': len(timed_lyrics),
            'duration': total_duration,
            'source': 'Whisper automatic alignment'
        },
        'lyrics': timed_lyrics
    }
    
    # Save timestamps.json
    os.makedirs(output_dir, exist_ok=True)
    output_file = os.path.join(output_dir, 'timestamps.json')
    
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)
    
    print(f'\n✅ Success!')
    print(f'   Aligned {len(timed_lyrics)} lyrics')
    print(f'   Saved to: {output_file}')
    print(f'\nNext steps:')
    print(f'  1. Generate images (if needed): npm start -- "{audio_file}" "{lyrics_file}" --output "{output_dir}"')
    print(f'  2. Generate video: npm run video "{audio_file}" "{output_dir}"')

def match_lyrics_to_segments(lyrics, segments):
    """Match user's lyrics to Whisper's detected segments"""
    timed_lyrics = []
    
    if not segments:
        print('   ⚠️  No speech detected by Whisper')
        return []
    
    # Strategy: Distribute segments evenly across lyrics
    segments_per_lyric = len(segments) / len(lyrics)
    
    for i, lyric in enumerate(lyrics):
        start_idx = int(i * segments_per_lyric)
        end_idx = int((i + 1) * segments_per_lyric)
        
        relevant_segments = segments[start_idx:end_idx]
        
        if relevant_segments:
            start_time = relevant_segments[0]['start']
            end_time = relevant_segments[-1]['end']
            
            timed_lyrics.append({
                'index': i,
                'startTime': start_time,
                'endTime': end_time,
                'duration': end_time - start_time,
                'text': lyric,
                'imagePath': f"images/lyric_{str(i).zfill(3)}.png"
            })
            
            preview = lyric[:40] + '...' if len(lyric) > 40 else lyric
            print(f'   {i + 1}. {start_time:.1f}s - {end_time:.1f}s: {preview}')
    
    return timed_lyrics

if __name__ == '__main__':
    main()

File: whisper-align.js
#!/usr/bin/env node

/**
 * Use Whisper to automatically detect lyric timing
 * Usage: node whisper-align.js <audio-file> <lyrics-file> <output-dir>
 */

const { exec } = require('child_process');
const { promisify } = require('util');
const fs = require('fs');
const path = require('path');

const execAsync = promisify(exec);

const audioFile = process.argv[2];
const lyricsFile = process.argv[3];
const outputDir = process.argv[4] || './output';

if (!audioFile || !lyricsFile) {
    console.log('Usage: node whisper-align.js <audio-file> <lyrics-file> [output-dir]');
    console.log('\nExample:');
    console.log('  node whisper-align.js song.mp3 lyrics.txt ./output');
    console.log('\nThis will:');
    console.log('  1. Use Whisper to detect when words are spoken');
    console.log('  2. Match detected words to your lyrics');
    console.log('  3. Create timestamps.json with accurate timing');
    process.exit(1);
}

async function main() {
    console.log('🎵 Whisper Automatic Alignment\n');
    
    // Read lyrics
    console.log('📖 Reading lyrics...');
    const lyrics = fs.readFileSync(lyricsFile, 'utf-8')
        .split('\n')
        .map(line => line.trim())
        .filter(line => line.length > 0);
    
    console.log(`   Found ${lyrics.length} lines\n`);
    
    // Run Whisper
    console.log('🎤 Running Whisper (this may take 2-5 minutes)...');
    console.log('   Detecting speech and timing...');
    
    const tempDir = path.join(outputDir, 'temp_whisper');
    if (!fs.existsSync(tempDir)) {
        fs.mkdirSync(tempDir, { recursive: true });
    }
    
    const audioName = path.basename(audioFile, path.extname(audioFile));
    
    try {
        // Run whisper with word-level timestamps
        const command = `whisper "${audioFile}" --model base --language hi --output_format json --output_dir "${tempDir}" --word_timestamps True`;
        
        await execAsync(command, { maxBuffer: 50 * 1024 * 1024 });
        
        console.log('   ✓ Whisper analysis complete\n');
        
        // Read Whisper output
        const whisperFile = path.join(tempDir, `${audioName}.json`);
        const whisperData = JSON.parse(fs.readFileSync(whisperFile, 'utf-8'));
        
        console.log('🔗 Matching lyrics to detected speech...');
        
        // Match lyrics to segments
        const timedLyrics = matchLyricsToSegments(lyrics, whisperData.segments);
        
        // Create timestamps.json
        const totalDuration = timedLyrics.length > 0 ? 
            timedLyrics[timedLyrics.length - 1].endTime : 0;
        
        const data = {
            version: '1.0',
            metadata: {
                generatedAt: new Date().toISOString(),
                totalLines: timedLyrics.length,
                duration: totalDuration,
                source: 'Whisper automatic alignment'
            },
            lyrics: timedLyrics
        };
        
        // Save timestamps.json
        if (!fs.existsSync(outputDir)) {
            fs.mkdirSync(outputDir, { recursive: true });
        }
        
        const outputFile = path.join(outputDir, 'timestamps.json');
        fs.writeFileSync(outputFile, JSON.stringify(data, null, 2));
        
        console.log(`\n✅ Success!`);
        console.log(`   Aligned ${timedLyrics.length} lyrics`);
        console.log(`   Saved to: ${outputFile}`);
        console.log(`\nNext steps:`);
        console.log(`  1. Generate images: npm start -- "${audioFile}" "${lyricsFile}" --output "${outputDir}"`);
        console.log(`  2. Or if images exist: npm run video "${audioFile}" "${outputDir}"`);
        
        // Clean up temp files
        // fs.rmSync(tempDir, { recursive: true, force: true });
        
    } catch (error) {
        console.error('\n❌ Error:', error.message);
        process.exit(1);
    }
}

/**
 * Match user's lyrics to Whisper's detected segments
 */
function matchLyricsToSegments(lyrics, segments) {
    const timedLyrics = [];
    
    if (segments.length === 0) {
        console.error('   ⚠️  No speech detected by Whisper');
        return [];
    }
    
    // Strategy: Distribute segments evenly across lyrics
    const segmentsPerLyric = segments.length / lyrics.length;
    
    for (let i = 0; i < lyrics.length; i++) {
        const startIdx = Math.floor(i * segmentsPerLyric);
        const endIdx = Math.floor((i + 1) * segmentsPerLyric);
        
        const relevantSegments = segments.slice(startIdx, endIdx);
        
        if (relevantSegments.length > 0) {
            const startTime = relevantSegments[0].start;
            const endTime = relevantSegments[relevantSegments.length - 1].end;
            
            timedLyrics.push({
                index: i,
                startTime,
                endTime,
                duration: endTime - startTime,
                text: lyrics[i],
                imagePath: `images/lyric_${String(i).padStart(3, '0')}.png`
            });
            
            console.log(`   ${i + 1}. ${startTime.toFixed(1)}s - ${endTime.toFixed(1)}s: ${lyrics[i].substring(0, 40)}...`);
        }
    }
    
    return timedLyrics;
}

main();
